mempool/src/validator.rs

```diff
@@ -0,0 +1,325 @@
+        }
+
+        Ok(())
+    }
+
+    fn validate_underpriced_tx_wo_chain_id(
+        chain_id: Option<u64>,
+        tx_gas_price: Option<u128>,
+        price: &GasPriceModel,
+    ) -> Result<(), PreFlightError> {
+        if chain_id.is_some() {
+            return Ok(());
+        }
+        if let Some(gas_price) = tx_gas_price {
+            let gas_price = U128::from(gas_price);
+            if gas_price < price.min_wo_chain_id_acceptable_gas_price {
+                return Err(PreFlightError::UnderpricedWoChainId);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+fn to_u256(v: alloy_primitives::U256) -> U256 {
+    let limbs = v.as_limbs();
+    let limb0 = (limbs[1] as u128) << 64 | (limbs[0] as u128); // Lower u128
+    let limb1 = (limbs[3] as u128) << 64 | (limbs[2] as u128); // Upper u128
+
+    U256([limb0, limb1])
+}
+
+fn tx_input_len(tx: &TxEnvelope) -> Result<usize, PreFlightError> {
+    match tx {
+        TxEnvelope::Legacy(signed) => Ok(signed.tx().input.len()),
+        TxEnvelope::Eip1559(signed) => Ok(signed.tx().input.len()),
+        TxEnvelope::Eip2930(signed) => Ok(signed.tx().input.len()),
+        TxEnvelope::Eip4844(signed) => Ok(signed.tx().tx().input.len()),
+        _ => Err(PreFlightError::UnsupportedTxType),
+    }
+}
+
+fn chain_id(tx: &TxEnvelope) -> Result<Option<u64>, PreFlightError> {
+    match tx {
+        TxEnvelope::Legacy(signed) => Ok(signed.tx().chain_id),
+        TxEnvelope::Eip1559(signed) => Ok(signed.tx().chain_id()),
+        TxEnvelope::Eip2930(signed) => Ok(signed.tx().chain_id()),
+        TxEnvelope::Eip4844(signed) => Ok(signed.tx().chain_id()),
+        _ => Err(PreFlightError::UnsupportedTxType),
+    }
+}
+
+fn tx_gas_limit(tx: &TxEnvelope) -> Result<u128, PreFlightError> {
+    match tx {
+        TxEnvelope::Legacy(signed) => Ok(signed.tx().gas_limit),
+        TxEnvelope::Eip1559(signed) => Ok(signed.tx().gas_limit()),
+        TxEnvelope::Eip2930(signed) => Ok(signed.tx().gas_limit()),
+        TxEnvelope::Eip4844(signed) => Ok(signed.tx().gas_limit()),
+        _ => Err(PreFlightError::UnsupportedTxType),
+    }
+}
+
+fn tx_gas_price(tx: &TxEnvelope) -> Result<Option<u128>, PreFlightError> {
+    match tx {
+        TxEnvelope::Legacy(signed) => Ok(signed.tx().gas_price()),
+        TxEnvelope::Eip1559(signed) => Ok(signed.tx().gas_price()),
+        TxEnvelope::Eip2930(signed) => Ok(signed.tx().gas_price()),
+        TxEnvelope::Eip4844(signed) => Ok(signed.tx().gas_price()),
+        _ => Err(PreFlightError::UnsupportedTxType),
+    }
+}
+
+fn tx_value(tx: &TxEnvelope) -> Result<alloy_primitives::U256, PreFlightError> {
+    match tx {
+        TxEnvelope::Legacy(signed) => Ok(signed.tx().value),
+        TxEnvelope::Eip1559(signed) => Ok(signed.tx().value),
+        TxEnvelope::Eip2930(signed) => Ok(signed.tx().value),
+        TxEnvelope::Eip4844(signed) => Ok(signed.tx().tx().value),
+        _ => Err(PreFlightError::UnsupportedTxType),
+    }
+}
```

Comment:
I think I've seen the same code while reviewing mempool stuff. Just add a `trait TxEnvelopeExt` to `common` or `proxy`.